### PR TITLE
#9095: Support for java.time.Duration

### LINKED ISSUE #9095: Support java.time.Duration in entity creation
##### **Overview of the feature request**

Add support for the java.time.Duration type as a field of an entity.

##### **Motivation for or Use Case**

Hibernate 5 natively supports this type, and little effort is necessary for supporting it in Mongo (a couple of converters are already defined in a companion JHipster project).

##### **Related issues or PR**

I did the homework but didn't find anything.

I am working on a PR for this. Help wanted for React! And feedback is welcome!

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by massimosiani**
use the duration to long converter instance

treat Duration as a number client-side

add a Duration field to the base test

add a missing import

consider the Duration field in the search criteria

Fix #9095

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by massimosiani** [comment]
This addresses #9095. I modified the app-sample-dev app to supplement it with test cases for a field of type java.time.Duration.

**Comment by DanielFran** [comment]
Do not forget that it will be needed PRs for jhipster-core and documentation.

**Comment by DanielFran** [comment]
Closing/Reopening to launching tests

**Comment by pvliss** [comment]
Closing and re-opening to relaunch CI tests

**Comment by MathieuAA** [comment]
I'm ready to merge the PR on JCore's end. How about you guys?

File: generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs [hunk 1]
```diff
@@ -109,6 +109,8 @@ describe('<%= entityClass %> e2e test', () => {
             <%= entityInstance %>UpdatePage.set<%= fieldNameCapitalized %>Input('2000-12-31'),
             <%_ } else if (['Instant', 'ZonedDateTime'].includes(fieldType)) { _%>
             <%= entityInstance %>UpdatePage.set<%= fieldNameCapitalized %>Input('01/01/2001' + protractor.Key.TAB + '02:30AM'),
+            <%_ } else if (['Duration'].includes(fieldType)) { _%>
```
Comment (MathieuAA) [inline generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: Just do `fieldType === 'Duration'`

File: generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs [hunk 2]
```diff
@@ -145,6 +147,8 @@ describe('<%= entityClass %> e2e test', () => {
         expect(await <%= entityInstance %>UpdatePage.get<%= fieldNameCapitalized %>Input()).to.eq('2000-12-31', 'Expected <%=fieldName%> value to be equals to 2000-12-31');
             <%_ } else if (['Instant', 'ZonedDateTime'].includes(fieldType)) { _%>
         expect(await <%= entityInstance %>UpdatePage.get<%= fieldNameCapitalized %>Input()).to.contain('2001-01-01T02:30', 'Expected <%=fieldName%> value to be equals to 2000-12-31');
+            <%_ } else if (['Duration'].includes(fieldType)) { _%>
```
Comment (MathieuAA) [inline generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: same

File: generators/entity-client/templates/react/src/test/javascript/e2e/entities/entity.spec.ts.ejs [hunk 3]
```diff
@@ -113,6 +113,9 @@ describe('<%= entityClass %> e2e test', () => {
             <%_ } else if (['Instant', 'ZonedDateTime'].includes(fieldType)) { _%>
             await <%= entityInstance %>UpdatePage.set<%= fieldNameCapitalized %>Input('01/01/2001' + protractor.Key.TAB + '02:30AM');
             expect(await <%= entityInstance %>UpdatePage.get<%= fieldNameCapitalized %>Input()).to.contain('2001-01-01T02:30');
+            <%_ } else if (['Duration'].includes(fieldType)) { _%>
```
Comment (MathieuAA) [inline generators/entity-client/templates/react/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: same

File: generators/entity-server/templates/src/main/java/package/common/search_template.ejs [hunk 4]
```diff
@@ -28,6 +28,7 @@
     <% } if (pagination !== 'no') { %>
     public ResponseEntity<List<<%= instanceType %>>> search<%= entityClassPlural %>(@RequestParam String query, Pageable pageable, HttpServletRequest request) {
         log.debug("REST request to search for a page of <%= entityClassPlural %> for query {}", query);<% if (viaService) { %>
+        pageable = Optional.ofNullable(pageable).orElse(Pageable.unpaged());
```
Comment (MathieuAA) [inline generators/entity-server/templates/src/main/java/package/common/search_template.ejs]: is this related to this PR?

Comment (massimosiani) [reply generators/entity-server/templates/src/main/java/package/common/search_template.ejs]: No, it isn't. Sorry, maybe my own modification for other purposes, and left there.
I rollbacked it.

File: generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs (hunk 1 above)
Comment (massimosiani) [reply generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: Sure!

File: generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs (hunk 2 above)
Comment (massimosiani) [reply generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: Done!

File: generators/entity-client/templates/react/src/test/javascript/e2e/entities/entity.spec.ts.ejs (hunk 3 above)
Comment (massimosiani) [reply generators/entity-client/templates/react/src/test/javascript/e2e/entities/entity.spec.ts.ejs]: Done again :)

File: test-integration/samples/.jhipster/BankAccount.json [hunk 5]
```diff
@@ -38,6 +38,10 @@
             "fieldName": "meanOperationDuration",
             "fieldType": "Double"
         },
+        {
+            "fieldName": "meanQueueDuration",
```
Comment (murdos) [inline test-integration/samples/.jhipster/BankAccount.json]: It would be better to add a field of type Duration in FieldTestEntity.json : it's used for nearly all database types, and already contains every field type that JHipster supports, so it would be incomplete without Duration ;)

Comment (massimosiani) [reply test-integration/samples/.jhipster/BankAccount.json]: Hi @murdos, I added it to the list of fields. Will the CI run the relative tests?

Comment (murdos) [reply test-integration/samples/.jhipster/BankAccount.json]: Thanks! Yes, it will.

**Comment by DanielFran** [comment]
Closing/Reopenning

**Comment by DanielFran** [comment]
@massimosiani Thanks for this PR!

@MathieuAA I am merging this, we will need soon a new version of jhipster-core.
